File: sdk/typescript/test/integration.test.ts
import { spawn, type ChildProcessWithoutNullStreams } from "node:child_process";
import fs from "node:fs/promises";
import { createServer } from "node:net";
import os from "node:os";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { afterAll, beforeAll, describe, expect, it } from "vitest";

import {
  TempliqxHttpError,
  TempliqxTransportError,
  createTempliqxClient,
  type TempliqxClient,
} from "../src/index.js";

const enabled = process.env.TEMPLIQX_SDK_IT === "1";
const repoRoot = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../../..");
const fingerprintPattern = /^(?:sha256:)?[a-f0-9]{64}$/;

async function freePort(): Promise<number> {
  return new Promise((resolve, reject) => {
    const server = createServer();
    server.once("error", reject);
    server.listen(0, "127.0.0.1", () => {
      const address = server.address();
      if (!address || typeof address === "string") {
        server.close();
        reject(new Error("Could not reserve a TCP port"));
        return;
      }
      server.close((error) => (error ? reject(error) : resolve(address.port)));
    });
  });
}

describe.skipIf(!enabled)("Templiqx SDK against the deterministic-fake HTTP server", () => {
  let server: ChildProcessWithoutNullStreams;
  let tempDir: string;
  let baseUrl: string;
  let client: TempliqxClient;
  let serverOutput = "";

  beforeAll(async () => {
    tempDir = await fs.mkdtemp(path.join(os.tmpdir(), "templiqx-sdk-it-"));
    const packagesRoot = path.join(tempDir, "packages");
    const workspaceRoot = path.join(tempDir, "workspace");
    await fs.mkdir(packagesRoot, { recursive: true });
    await fs.mkdir(workspaceRoot, { recursive: true });
    const port = await freePort();
    baseUrl = `http://127.0.0.1:${port}`;
    const { MODEL_API_KEY: _modelApiKey, ...environment } = process.env;

    server = spawn("cargo", ["run", "--quiet", "-p", "templiqx-http-server"], {
      cwd: repoRoot,
      env: {
        ...environment,
        TEMPLIQX_HTTP_ADDR: `127.0.0.1:${port}`,
        TEMPLIQX_ROOT: packagesRoot,
        TEMPLIQX_WORKSPACE: workspaceRoot,
      },
      stdio: ["pipe", "pipe", "pipe"],
    });
    server.stdout.on("data", (chunk) => (serverOutput += chunk.toString()));
    server.stderr.on("data", (chunk) => (serverOutput += chunk.toString()));

    const deadline = Date.now() + 120_000;
    while (Date.now() < deadline) {
      if (server.exitCode !== null) throw new Error(`Server exited during startup:\n${serverOutput}`);
      try {
        const response = await fetch(`${baseUrl}/operations/v1/health/ready`);
        if (response.ok) break;
      } catch {
        // The server is still compiling or binding its listener.
      }
      await new Promise((resolve) => setTimeout(resolve, 100));
    }
    const ready = await fetch(`${baseUrl}/operations/v1/health/ready`).catch(() => undefined);
    if (!ready?.ok) throw new Error(`Server did not become ready:\n${serverOutput}`);
    client = createTempliqxClient({ baseUrl, timeoutMs: 5_000 });
  }, 130_000);

  afterAll(async () => {
    if (server && server.exitCode === null) {
      server.kill("SIGTERM");
      await new Promise<void>((resolve) => {
        server.once("exit", () => resolve());
        setTimeout(resolve, 5_000);
      });
    }
    if (tempDir) await fs.rm(tempDir, { recursive: true, force: true });
  });

  it("drives health, catalog, CAS, compile, execute, HTTP error, and abort paths", async () => {
    const live = await client.getOperationsV1Liveness();
    const ready = await client.getOperationsV1Readiness();
    expect(live.data.status).toBe("ok");
    expect(ready.data.status).toBe("ready");

    const catalog = await client.catalog();
    expect(catalog.data.ok).toBe(true);
    expect(catalog.data.result).toContain("execute_contract");

    const created = await client.createPackage({ name: "sdk-it", version: "0.1.0" });
    const packageFingerprint = created.data.fingerprints.package;
    expect(packageFingerprint).toMatch(fingerprintPattern);
    const updated = await client.updatePackage(
      { package: "sdk-it" },
      { description: "TypeScript SDK integration" },
      { ifMatch: packageFingerprint! },
    );
    expect(updated.data.ok).toBe(true);

    const contractSource = await fs.readFile(
      path.join(repoRoot, "examples/packages/demo/contracts/greeting.yaml"),
      "utf8",
    );
    const put = await client.putContract(
      { package: "sdk-it", contract: "greeting" },
      contractSource,
    );
    expect(put.data.ok).toBe(true);

    const render = {
      inputs: { name: "Ryan" },
      context: { organization: "Blinqx" },
    };
    const compile = await client.compileContract(
      { package: "sdk-it", contract: "greeting" },
      { render, capabilities: ["structured_output"] },
    );
    expect(compile.data.ok).toBe(true);

    const execution = await client.executeContract(
      { package: "sdk-it", contract: "greeting" },
      {
        render,
        capabilities: ["structured_output"],
        fixture_output: { greeting: "Hello Ryan" },
        stream: false,
      },
    );
    expect(execution.data.ok).toBe(true);
    expect(execution.data.result?.output_fingerprint).toMatch(fingerprintPattern);
    console.log(`ExecutionReceipt fingerprint: ${execution.data.result!.output_fingerprint}`);

    const httpError = await client
      .inspectContract({ package: "missing", contract: "greeting" })
      .catch((error: unknown) => error);
    expect(httpError).toBeInstanceOf(TempliqxHttpError);
    expect(httpError).toMatchObject({ status: 404 });
    expect((httpError as TempliqxHttpError).envelope?.diagnostics[0]?.code).toBe("TQX_NOT_FOUND");

    const controller = new AbortController();
    controller.abort(new Error("integration cancellation"));
    await expect(
      client.catalog({ signal: controller.signal, requestId: "sdk-it-abort" }),
    ).rejects.toBeInstanceOf(TempliqxTransportError);
  }, 30_000);
});

File: sdk/go/integration_test.go
//go:build integration

package templiqx

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"net"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"strings"
	"syscall"
	"testing"
	"time"
)

var fingerprintPattern = regexp.MustCompile(`^(?:sha256:)?[a-f0-9]{64}$`)

func TestIntegrationAgainstRealServer(t *testing.T) {
	if os.Getenv("TEMPLIQX_SDK_IT") != "1" {
		t.Skip("set TEMPLIQX_SDK_IT=1 to run the live-server integration test")
	}

	repoRoot, err := filepath.Abs(filepath.Join("..", ".."))
	if err != nil {
		t.Fatalf("repo root: %v", err)
	}
	tempDir := t.TempDir()
	packagesRoot := filepath.Join(tempDir, "packages")
	workspaceRoot := filepath.Join(tempDir, "workspace")
	for _, directory := range []string{packagesRoot, workspaceRoot} {
		if err := os.MkdirAll(directory, 0o755); err != nil {
			t.Fatalf("create %s: %v", directory, err)
		}
	}
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		t.Fatalf("reserve port: %v", err)
	}
	port := listener.Addr().(*net.TCPAddr).Port
	if err := listener.Close(); err != nil {
		t.Fatalf("release port: %v", err)
	}
	baseURL := fmt.Sprintf("http://127.0.0.1:%d", port)

	command := exec.Command("cargo", "run", "--quiet", "-p", "templiqx-http-server")
	command.Dir = repoRoot
	command.Env = append(filteredEnvironment(
		"MODEL_API_KEY", "TEMPLIQX_HTTP_ADDR", "TEMPLIQX_ROOT", "TEMPLIQX_WORKSPACE", "TEMPLIQX_RUNTIME_MODE",
	),
		fmt.Sprintf("TEMPLIQX_HTTP_ADDR=127.0.0.1:%d", port),
		"TEMPLIQX_ROOT="+packagesRoot,
		"TEMPLIQX_WORKSPACE="+workspaceRoot,
		"TEMPLIQX_RUNTIME_MODE=deterministic-fake",
	)
	var serverOutput bytes.Buffer
	command.Stdout = &serverOutput
	command.Stderr = &serverOutput
	if err := command.Start(); err != nil {
		t.Fatalf("start server: %v", err)
	}
	exit := make(chan error, 1)
	go func() { exit <- command.Wait() }()
	serverExited := false
	t.Cleanup(func() {
		if serverExited || command.Process == nil {
			return
		}
		_ = command.Process.Signal(syscall.SIGTERM)
		select {
		case <-exit:
		case <-time.After(5 * time.Second):
			_ = command.Process.Kill()
			<-exit
		}
	})

	readinessClient := &http.Client{Timeout: 500 * time.Millisecond}
	deadline := time.Now().Add(120 * time.Second)
	ready := false
	for time.Now().Before(deadline) {
		select {
		case err := <-exit:
			serverExited = true
			t.Fatalf("server exited during startup: %v\n%s", err, serverOutput.String())
		default:
		}
		response, requestErr := readinessClient.Get(baseURL + "/operations/v1/health/ready")
		if requestErr == nil {
			_ = response.Body.Close()
			if response.StatusCode >= http.StatusOK && response.StatusCode < http.StatusMultipleChoices {
				ready = true
				break
			}
		}
		time.Sleep(100 * time.Millisecond)
	}
	if !ready {
		t.Fatalf("server did not become ready:\n%s", serverOutput.String())
	}

	client, err := NewClient(baseURL, WithClientTimeout(5*time.Second))
	if err != nil {
		t.Fatalf("NewClient: %v", err)
	}
	ctx := context.Background()
	live, err := client.GetOperationsV1Liveness(ctx)
	if err != nil || live.Data.Status != Ok {
		t.Fatalf("liveness = %#v, %v", live, err)
	}
	readyStatus, err := client.GetOperationsV1Readiness(ctx)
	if err != nil || readyStatus.Data.Status != Ready {
		t.Fatalf("readiness = %#v, %v", readyStatus, err)
	}
	catalog, err := client.Catalog(ctx)
	if err != nil || !catalog.Data.Ok || catalog.Data.Result == nil || !contains(*catalog.Data.Result, "execute_contract") {
		t.Fatalf("catalog = %#v, %v", catalog, err)
	}

	created, err := client.CreatePackage(ctx, CreatePackageRequest{Name: "sdk-go-it", Version: "0.1.0"})
	if err != nil {
		t.Fatalf("CreatePackage: %v", err)
	}
	packageFingerprint := created.Data.Fingerprints["package"]
	if !fingerprintPattern.MatchString(packageFingerprint) {
		t.Fatalf("package fingerprint = %q", packageFingerprint)
	}
	description := "Go SDK integration"
	if _, err := client.UpdatePackage(
		ctx, "sdk-go-it", UpdatePackageRequest{Description: &description}, WithIfMatch(packageFingerprint),
	); err != nil {
		t.Fatalf("UpdatePackage: %v", err)
	}

	contractSource, err := os.ReadFile(filepath.Join(repoRoot, "examples", "packages", "demo", "contracts", "greeting.yaml"))
	if err != nil {
		t.Fatalf("read greeting contract: %v", err)
	}
	if _, err := client.PutContract(ctx, "sdk-go-it", "greeting", string(contractSource)); err != nil {
		t.Fatalf("PutContract: %v", err)
	}
	inputs := map[string]JsonValue{"name": "Ryan"}
	contractContext := map[string]JsonValue{"organization": "Blinqx"}
	capabilities := []string{"structured_output"}
	render := &RenderRequest{Inputs: &inputs, Context: &contractContext}
	compiled, err := client.CompileContract(
		ctx, "sdk-go-it", "greeting", CompileRequest{Render: render, Capabilities: &capabilities},
	)
	if err != nil || !compiled.Data.Ok {
		t.Fatalf("CompileContract = %#v, %v", compiled, err)
	}
	fixtureOutput := JsonValue(map[string]any{"greeting": "Hello Ryan"})
	stream := false
	executed, err := client.ExecuteContract(ctx, "sdk-go-it", "greeting", ExecuteRequest{
		Render: render, Capabilities: &capabilities, FixtureOutput: &fixtureOutput, Stream: &stream,
	})
	if err != nil || !executed.Data.Ok || executed.Data.Result == nil {
		t.Fatalf("ExecuteContract = %#v, %v", executed, err)
	}
	fingerprint := executed.Data.Result.OutputFingerprint
	if !fingerprintPattern.MatchString(fingerprint) {
		t.Fatalf("ExecutionReceipt output fingerprint = %q", fingerprint)
	}
	t.Logf("ExecutionReceipt fingerprint: %s", fingerprint)

	_, err = client.InspectContract(ctx, "missing", "greeting")
	var httpError *TempliqxHTTPError
	if !errors.As(err, &httpError) || httpError.StatusCode != http.StatusNotFound ||
		httpError.Envelope == nil || httpError.Envelope.Diagnostics[0].Code != "TQX_NOT_FOUND" {
		t.Fatalf("missing contract error = %#v", err)
	}
	cancelled, cancel := context.WithCancel(ctx)
	cancel()
	_, err = client.Catalog(cancelled, WithRequestID("sdk-go-it-abort"))
	var transportError *TempliqxTransportError
	if !errors.As(err, &transportError) {
		t.Fatalf("cancel error = %#v", err)
	}
}

func filteredEnvironment(excluded ...string) []string {
	prefixes := make([]string, len(excluded))
	for index, name := range excluded {
		prefixes[index] = name + "="
	}
	result := make([]string, 0, len(os.Environ()))
	for _, entry := range os.Environ() {
		keep := true
		for _, prefix := range prefixes {
			if strings.HasPrefix(entry, prefix) {
				keep = false
				break
			}
		}
		if keep {
			result = append(result, entry)
		}
	}
	return result
}

func contains(values []string, expected string) bool {
	for _, value := range values {
		if value == expected {
			return true
		}
	}
	return false
}
